scrapy/settings/__init__.py

```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
 
     def set(self, value, priority):
         """Sets value if priority is higher or equal than current priority."""
-        if priority >= self.priority:
-            self.value = value
-            self.priority = priority
+        if isinstance(self.value, BaseSettings) and isinstance(value, Mapping):
+            # Ignore self.priority if self.value has per-key priorities
+            # XXX: Second condition allows completely replacing a BaseSettings
+            #      instance with a value of incompatible type (i.e. something
+            #      that does not map keys to values, e.g.a list). I included
+            #      this b/c the test_cmdline/__init__.py test uses a deprecated
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
```

Comment:
The thing is that the `EXTENSIONS` test was only an example of what developers might actually *want* to do. What if they have a settings that is a `BaseSettings` instance, but for some reason they decide to replace it with a different kind of variable, say `None`. The `isinstance(value, Mapping)` allows for this.
I guess the alternative is to make developers actively acknowledge that they're losing the `BaseSettings` instance and all its values/priorites by deleting it, then inserting a new setting with the type they want. While that is more explicit, it's probably not expected behaviour, given that "type replacement" for all other types of settings does not yield exceptions.